eta/core/c3d.py

```diff
@@ -0,0 +1,381 @@
+'''
+TensorFlow implementation of the popular C3D network.
+
+This module extends and add functionality to the original C3D implementation
+by Hou Xin.
+
+C3D implementation in TensorFlow:
+https://github.com/hx173149/C3D-tensorflow
+Hou Xin, 2016
+
+Copyright 2018, Voxel51, LLC
+voxel51.com
+
+Yixin Jin, yixin@voxel51.com
+Brian Moore, brian@voxel51.com
+'''
+# pragma pylint: disable=redefined-builtin
+# pragma pylint: disable=unused-wildcard-import
+# pragma pylint: disable=wildcard-import
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+from builtins import *
+# pragma pylint: enable=redefined-builtin
+# pragma pylint: enable=unused-wildcard-import
+# pragma pylint: enable=wildcard-import
+
+import logging
+import math
+import os
+
+import numpy as np
+from sklearn.preprocessing import normalize
+import tensorflow as tf
+
+from eta.core.config import Config
+import eta.core.image as etai
+from eta.core.features import Featurizer
+from eta.core.tfutils import TensorFlowModelWeights
+import eta.core.video as etav
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_first_k_frames(inpath, k, embedding_frame_size):
+    '''Sample first k frames in a video. Return a numpy array of size
+    [k, embedding_frame_siz, embedding_frame_size]
+
+    Args:
+        inpath: path to the input video
+        k: number of frames to extract for the input video
+        embedding_frame_size: size of the output frames
+    '''
+    data = []
+    num_frames = etav.get_frame_count(inpath)
+    assert k < num_frames
+    with etav.FFmpegVideoReader(inpath, "1-%d" % k) as vr:
+        for img in vr:
+            img = etai.resize(img, embedding_frame_size, embedding_frame_size)
+            data.append(img)
+    np_arr_data = np.array(data).astype(np.float32)
+    return np_arr_data
+
+
+def uniformly_sample_k_frames(inpath, k, embedding_frame_size):
+    '''Uniformly sample k frames, always including the first frame. Return
+    a numpy array of size [k, embedding_frame_siz, embedding_frame_size]
+
+    Args:
+        inpath: path to the input video
+        k: number of frames to extract for the input video
+        embedding_frame_size: dimension to use when embedding the frames
+    '''
+    data = []
+    num_frames = etav.get_frame_count(inpath)
+    assert k < num_frames
```

Comment:
you can write this whole mess as:
```py
rng = [int(round(i)) for i in np.linspace(1, num_frames, k)]
```
and it will even work in a reasonable way if there aren't enough frames

Reply:
Nice!